File: sdk/python/snoopscan/_update.py
"""A quiet "a newer version is available" line, as every comparable CLI has.

People install a CLI once and run it for months; without a nudge a login flow
or a fixed bug never reaches them. The check asks PyPI at most once a day,
shows the line at most twice a day, goes to stderr so it never mixes with a
command's output, and can never make a command fail. SNOOPSCAN_NO_UPDATE_CHECK=1
turns it off.
"""

from __future__ import annotations

import json
import os
import time
from pathlib import Path

import httpx

from . import config as user_config

PYPI_URL = "https://pypi.org/pypi/snoopscan/json"
CHECK_EVERY_S = 20 * 3600
SHOW_EVERY_S = 12 * 3600
DISABLE_ENV = "SNOOPSCAN_NO_UPDATE_CHECK"


def _cache_path() -> Path:
    return user_config.config_path().parent / "update-check.json"


def _parse(version: str) -> tuple[int, ...]:
    out = []
    for part in version.split("."):
        digits = "".join(ch for ch in part if ch.isdigit())
        out.append(int(digits) if digits else 0)
    return tuple(out)


def is_newer(latest: str, current: str) -> bool:
    return _parse(latest) > _parse(current)


def disabled() -> bool:
    return os.environ.get(DISABLE_ENV, "").lower() in ("1", "true", "yes")


def _read() -> dict[str, object]:
    try:
        data = json.loads(_cache_path().read_text(encoding="utf-8"))
        return data if isinstance(data, dict) else {}
    except (OSError, ValueError):
        return {}


def _write(data: dict[str, object]) -> None:
    try:
        path = _cache_path()
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(json.dumps(data), encoding="utf-8")
    except OSError:
        pass  # a read-only home must not break a command


def latest_version(now: float | None = None, *, force: bool = False) -> str | None:
    """PyPI's newest version, from a day-old cache when there is one."""
    now = time.time() if now is None else now
    cache = _read()
    checked = cache.get("checked_at")
    if not force and isinstance(checked, (int, float)) and now - checked < CHECK_EVERY_S:
        latest = cache.get("latest")
        return latest if isinstance(latest, str) else None
    try:
        latest = str(httpx.get(PYPI_URL, timeout=2.0).json()["info"]["version"])
    except Exception:  # noqa: BLE001 - offline, blocked, slow: no notice, no error
        return None
    _write({**cache, "latest": latest, "checked_at": now})
    return latest


def notice(current: str, now: float | None = None) -> str | None:
    """The line to print, or None. Shown again only after a while, or for a newer release."""
    if disabled() or current in ("", "0+unknown"):
        return None
    now = time.time() if now is None else now
    latest = latest_version(now)
    if not latest or not is_newer(latest, current):
        return None
    cache = _read()
    shown_at = cache.get("shown_at")
    if (
        cache.get("shown_version") == latest
        and isinstance(shown_at, (int, float))
        and now - shown_at < SHOW_EVERY_S
    ):
        return None
    _write({**cache, "shown_version": latest, "shown_at": now})
    return (
        f"A newer snoopscan is available: {current} -> {latest}\n"
        "  Update: pipx upgrade snoopscan  |  uv tool upgrade snoopscan  |  "
        "python3 -m pip install -U snoopscan\n"
        f"  (npx and uvx always run the latest.)  Silence this: {DISABLE_ENV}=1"
    )

File: sdk/js/src/update.ts
/**
 * A quiet "a newer version is available" line, as every comparable CLI has —
 * the JS twin of the Python CLI's `_update.py`. Asks the npm registry at most
 * once a day, shows the line at most twice a day, writes to stderr so it never
 * mixes with a command's output, and can never make a command fail.
 * SNOOPSCAN_NO_UPDATE_CHECK=1 turns it off.
 */
import { mkdirSync, readFileSync, writeFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { configPath } from './config.js';

export const REGISTRY_URL = 'https://registry.npmjs.org/snoopscan/latest';
const CHECK_EVERY_MS = 20 * 3600 * 1000;
const SHOW_EVERY_MS = 12 * 3600 * 1000;
export const DISABLE_ENV = 'SNOOPSCAN_NO_UPDATE_CHECK';

interface Cache {
  latest?: string;
  checkedAt?: number;
  shownVersion?: string;
  shownAt?: number;
}

/** This package's own version, read from the package.json beside the build. */
export function currentVersion(): string {
  try {
    return String(JSON.parse(readFileSync(new URL('../package.json', import.meta.url), 'utf8')).version ?? '');
  } catch {
    return '';
  }
}

function parse(v: string): number[] {
  return v.split('.').map((p) => Number(p.replace(/\D/g, '')) || 0);
}

export function isNewer(latest: string, current: string): boolean {
  const a = parse(latest);
  const b = parse(current);
  for (let i = 0; i < Math.max(a.length, b.length); i++) {
    if ((a[i] ?? 0) !== (b[i] ?? 0)) return (a[i] ?? 0) > (b[i] ?? 0);
  }
  return false;
}

export function disabled(): boolean {
  return ['1', 'true', 'yes'].includes(String(process.env[DISABLE_ENV] ?? '').toLowerCase());
}

const cachePath = (): string => join(dirname(configPath()), 'update-check.json');

function read(): Cache {
  try {
    return JSON.parse(readFileSync(cachePath(), 'utf8')) as Cache;
  } catch {
    return {};
  }
}

function write(cache: Cache): void {
  try {
    mkdirSync(dirname(cachePath()), { recursive: true });
    writeFileSync(cachePath(), JSON.stringify(cache));
  } catch {
    // a read-only home must not break a command
  }
}

export async function latestVersion(
  opts: { now?: number; force?: boolean; fetchImpl?: typeof fetch } = {},
): Promise<string | undefined> {
  const now = opts.now ?? Date.now();
  const cache = read();
  if (!opts.force && typeof cache.checkedAt === 'number' && now - cache.checkedAt < CHECK_EVERY_MS) {
    return cache.latest;
  }
  try {
    const res = await (opts.fetchImpl ?? fetch)(REGISTRY_URL, { signal: AbortSignal.timeout(2000) });
    const latest = String(((await res.json()) as { version?: string }).version ?? '');
    if (!latest) return undefined;
    write({ ...cache, latest, checkedAt: now });
    return latest;
  } catch {
    return undefined; // offline, blocked, slow: no notice, no error
  }
}

/** The line to print, or undefined. Shown again only after a while, or for a newer release. */
export async function notice(
  current: string,
  opts: { now?: number; fetchImpl?: typeof fetch } = {},
): Promise<string | undefined> {
  if (disabled() || !current) return undefined;
  const now = opts.now ?? Date.now();
  const latest = await latestVersion({ now, fetchImpl: opts.fetchImpl });
  if (!latest || !isNewer(latest, current)) return undefined;
  const cache = read();
  if (cache.shownVersion === latest && typeof cache.shownAt === 'number' && now - cache.shownAt < SHOW_EVERY_MS) {
    return undefined;
  }
  write({ ...cache, shownVersion: latest, shownAt: now });
  return (
    `A newer snoopscan is available: ${current} -> ${latest}\n` +
    '  Update: npm install -g snoopscan@latest   (npx snoopscan@latest always runs the latest)\n' +
    `  Silence this: ${DISABLE_ENV}=1\n`
  );
}
